File: plugins/examples/discord-rich-presence/rpc.go
// Discord Rich Presence Plugin - RPC Communication
//
// This file handles all Discord gateway communication including WebSocket connections,
// presence updates, and heartbeat management. The discordRPC struct implements WebSocket
// callback interfaces and encapsulates all Discord communication logic.
package main

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/navidrome/navidrome/plugins/pdk/go/host"
	"github.com/navidrome/navidrome/plugins/pdk/go/pdk"
	"github.com/navidrome/navidrome/plugins/pdk/go/websocket"
)

// Discord WebSocket Gateway constants
const (
	heartbeatOpCode = 1 // Heartbeat operation code
	gateOpCode      = 2 // Identify operation code
	presenceOpCode  = 3 // Presence update operation code
)

const (
	heartbeatInterval = 41 // Heartbeat interval in seconds
	defaultImage      = "https://i.imgur.com/hb3XPzA.png"
)

// Scheduler callback payloads for routing
const (
	payloadHeartbeat     = "heartbeat"
	payloadClearActivity = "clear-activity"
)

// discordRPC handles Discord gateway communication and implements WebSocket callbacks.
type discordRPC struct{}

// ============================================================================
// WebSocket Callback Implementation
// ============================================================================

// OnTextMessage handles incoming WebSocket text messages.
func (r *discordRPC) OnTextMessage(input websocket.OnTextMessageRequest) error {
	return r.handleWebSocketMessage(input.ConnectionID, input.Message)
}

// OnBinaryMessage handles incoming WebSocket binary messages.
func (r *discordRPC) OnBinaryMessage(input websocket.OnBinaryMessageRequest) error {
	pdk.Log(pdk.LogDebug, fmt.Sprintf("Received unexpected binary message for connection '%s'", input.ConnectionID))
	return nil
}

// OnError handles WebSocket errors.
func (r *discordRPC) OnError(input websocket.OnErrorRequest) error {
	pdk.Log(pdk.LogWarn, fmt.Sprintf("WebSocket error for connection '%s': %s", input.ConnectionID, input.Error))
	return nil
}

// OnClose handles WebSocket connection closure.
func (r *discordRPC) OnClose(input websocket.OnCloseRequest) error {
	pdk.Log(pdk.LogInfo, fmt.Sprintf("WebSocket connection '%s' closed with code %d: %s", input.ConnectionID, input.Code, input.Reason))
	return nil
}

// activity represents a Discord activity.
type activity struct {
	Name        string             `json:"name"`
	Type        int                `json:"type"`
	Details     string             `json:"details"`
	State       string             `json:"state"`
	Application string             `json:"application_id"`
	Timestamps  activityTimestamps `json:"timestamps"`
	Assets      activityAssets     `json:"assets"`
}

type activityTimestamps struct {
	Start int64 `json:"start"`
	End   int64 `json:"end"`
}

type activityAssets struct {
	LargeImage string `json:"large_image"`
	LargeText  string `json:"large_text"`
}

// presencePayload represents a Discord presence update.
type presencePayload struct {
	Activities []activity `json:"activities"`
	Since      int64      `json:"since"`
	Status     string     `json:"status"`
	Afk        bool       `json:"afk"`
}

// identifyPayload represents a Discord identify payload.
type identifyPayload struct {
	Token      string             `json:"token"`
	Intents    int                `json:"intents"`
	Properties identifyProperties `json:"properties"`
}

type identifyProperties struct {
	OS      string `json:"os"`
	Browser string `json:"browser"`
	Device  string `json:"device"`
}

// ============================================================================
// Image Processing
// ============================================================================

// processImage processes an image URL for Discord, with fallback to default image.
func (r *discordRPC) processImage(imageURL, clientID, token string, isDefaultImage bool) (string, error) {
	if imageURL == "" {
		if isDefaultImage {
			return "", fmt.Errorf("default image URL is empty")
		}
		return r.processImage(defaultImage, clientID, token, true)
	}

	if strings.HasPrefix(imageURL, "mp:") {
		return imageURL, nil
	}

	// Check cache first
	cacheKey := fmt.Sprintf("discord.image.%x", imageURL)
	cachedValue, exists, err := host.CacheGetString(cacheKey)
	if err == nil && exists {
		pdk.Log(pdk.LogDebug, fmt.Sprintf("Cache hit for image URL: %s", imageURL))
		return cachedValue, nil
	}

	// Process via Discord API
	body := fmt.Sprintf(`{"urls":[%q]}`, imageURL)
	req := pdk.NewHTTPRequest(pdk.MethodPost, fmt.Sprintf("https://discord.com/api/v9/applications/%s/external-assets", clientID))
	req.SetHeader("Authorization", token)
	req.SetHeader("Content-Type", "application/json")
	req.SetBody([]byte(body))

	resp := req.Send()
	if resp.Status() >= 400 {
		if isDefaultImage {
			return "", fmt.Errorf("failed to process default image: HTTP %d", resp.Status())
		}
		return r.processImage(defaultImage, clientID, token, true)
	}

	var data []map[string]string
	if err := json.Unmarshal(resp.Body(), &data); err != nil {
		if isDefaultImage {
			return "", fmt.Errorf("failed to unmarshal default image response: %w", err)
		}
		return r.processImage(defaultImage, clientID, token, true)
	}

	if len(data) == 0 {
		if isDefaultImage {
			return "", fmt.Errorf("no data returned for default image")
		}
		return r.processImage(defaultImage, clientID, token, true)
	}

	image := data[0]["external_asset_path"]
	if image == "" {
		if isDefaultImage {
			return "", fmt.Errorf("empty external_asset_path for default image")
		}
		return r.processImage(defaultImage, clientID, token, true)
	}

	processedImage := fmt.Sprintf("mp:%s", image)

	// Cache the processed image URL
	var ttl int64 = 4 * 60 * 60 // 4 hours for regular images
	if isDefaultImage {
		ttl = 48 * 60 * 60 // 48 hours for default image
	}

	_ = host.CacheSetString(cacheKey, processedImage, ttl)
	pdk.Log(pdk.LogDebug, fmt.Sprintf("Cached processed image URL for %s (TTL: %ds)", imageURL, ttl))

	return processedImage, nil
}

// ============================================================================
// Activity Management
// ============================================================================

// sendActivity sends an activity update to Discord.
func (r *discordRPC) sendActivity(clientID, username, token string, data activity) error {
	pdk.Log(pdk.LogInfo, fmt.Sprintf("Sending activity for user %s: %s - %s", username, data.Details, data.State))

	processedImage, err := r.processImage(data.Assets.LargeImage, clientID, token, false)
	if err != nil {
		pdk.Log(pdk.LogWarn, fmt.Sprintf("Failed to process image for user %s, continuing without image: %v", username, err))
		data.Assets.LargeImage = ""
	} else {
		data.Assets.LargeImage = processedImage
	}

	presence := presencePayload{
		Activities: []activity{data},
		Status:     "dnd",
		Afk:        false,
	}
	return r.sendMessage(username, presenceOpCode, presence)
}

// clearActivity clears the Discord activity for a user.
func (r *discordRPC) clearActivity(username string) error {
	pdk.Log(pdk.LogInfo, fmt.Sprintf("Clearing activity for user %s", username))
	return r.sendMessage(username, presenceOpCode, presencePayload{})
}

// ============================================================================
// Low-level Communication
// ============================================================================

// sendMessage sends a message over the WebSocket connection.
func (r *discordRPC) sendMessage(username string, opCode int, payload any) error {
	message := map[string]any{
		"op": opCode,
		"d":  payload,
	}
	b, err := json.Marshal(message)
	if err != nil {
		return fmt.Errorf("failed to marshal message: %w", err)
	}

	err = host.WebSocketSendText(username, string(b))
	if err != nil {
		return fmt.Errorf("failed to send message: %w", err)
	}
	return nil
}

// getDiscordGateway retrieves the Discord gateway URL.
func (r *discordRPC) getDiscordGateway() (string, error) {
	req := pdk.NewHTTPRequest(pdk.MethodGet, "https://discord.com/api/gateway")
	resp := req.Send()
	if resp.Status() != 200 {
		return "", fmt.Errorf("failed to get Discord gateway: HTTP %d", resp.Status())
	}

	var result map[string]string
	if err := json.Unmarshal(resp.Body(), &result); err != nil {
		return "", fmt.Errorf("failed to parse Discord gateway response: %w", err)
	}
	return result["url"], nil
}

// sendHeartbeat sends a heartbeat to Discord.
func (r *discordRPC) sendHeartbeat(username string) error {
	seqNum, _, err := host.CacheGetInt(fmt.Sprintf("discord.seq.%s", username))
	if err != nil {
		return fmt.Errorf("failed to get sequence number: %w", err)
	}

	pdk.Log(pdk.LogDebug, fmt.Sprintf("Sending heartbeat for user %s: %d", username, seqNum))
	return r.sendMessage(username, heartbeatOpCode, seqNum)
}

// cleanupFailedConnection cleans up a failed Discord connection.
func (r *discordRPC) cleanupFailedConnection(username string) {
	pdk.Log(pdk.LogInfo, fmt.Sprintf("Cleaning up failed connection for user %s", username))

	// Cancel the heartbeat schedule
	if err := host.SchedulerCancelSchedule(username); err != nil {
		pdk.Log(pdk.LogWarn, fmt.Sprintf("Failed to cancel heartbeat schedule for user %s: %v", username, err))
	}

	// Close the WebSocket connection
	if err := host.WebSocketCloseConnection(username, 1000, "Connection lost"); err != nil {
		pdk.Log(pdk.LogWarn, fmt.Sprintf("Failed to close WebSocket connection for user %s: %v", username, err))
	}

	// Clean up cache entries
	_ = host.CacheRemove(fmt.Sprintf("discord.seq.%s", username))

	pdk.Log(pdk.LogInfo, fmt.Sprintf("Cleaned up connection for user %s", username))
}

// isConnected checks if a user is connected to Discord by testing the heartbeat.
func (r *discordRPC) isConnected(username string) bool {
	err := r.sendHeartbeat(username)
	if err != nil {
		pdk.Log(pdk.LogDebug, fmt.Sprintf("Heartbeat test failed for user %s: %v", username, err))
		return false
	}
	return true
}

// connect establishes a connection to Discord for a user.
func (r *discordRPC) connect(username, token string) error {
	if r.isConnected(username) {
		pdk.Log(pdk.LogInfo, fmt.Sprintf("Reusing existing connection for user %s", username))
		return nil
	}
	pdk.Log(pdk.LogInfo, fmt.Sprintf("Creating new connection for user %s", username))

	// Get Discord Gateway URL
	gateway, err := r.getDiscordGateway()
	if err != nil {
		return fmt.Errorf("failed to get Discord gateway: %w", err)
	}
	pdk.Log(pdk.LogDebug, fmt.Sprintf("Using gateway: %s", gateway))

	// Connect to Discord Gateway
	_, err = host.WebSocketConnect(gateway, nil, username)
	if err != nil {
		return fmt.Errorf("failed to connect to WebSocket: %w", err)
	}

	// Send identify payload
	payload := identifyPayload{
		Token:   token,
		Intents: 0,
		Properties: identifyProperties{
			OS:      "Windows 10",
			Browser: "Discord Client",
			Device:  "Discord Client",
		},
	}
	if err := r.sendMessage(username, gateOpCode, payload); err != nil {
		return fmt.Errorf("failed to send identify payload: %w", err)
	}

	// Schedule heartbeats for this user/connection
	cronExpr := fmt.Sprintf("@every %ds", heartbeatInterval)
	scheduleID, err := host.SchedulerScheduleRecurring(cronExpr, payloadHeartbeat, username)
	if err != nil {
		return fmt.Errorf("failed to schedule heartbeat: %w", err)
	}
	pdk.Log(pdk.LogInfo, fmt.Sprintf("Scheduled heartbeat for user %s with ID %s", username, scheduleID))

	pdk.Log(pdk.LogInfo, fmt.Sprintf("Successfully authenticated user %s", username))
	return nil
}

// disconnect closes the Discord connection for a user.
func (r *discordRPC) disconnect(username string) error {
	if err := host.SchedulerCancelSchedule(username); err != nil {
		return fmt.Errorf("failed to cancel schedule: %w", err)
	}

	if err := host.WebSocketCloseConnection(username, 1000, "Navidrome disconnect"); err != nil {
		return fmt.Errorf("failed to close WebSocket connection: %w", err)
	}
	return nil
}

// handleWebSocketMessage processes incoming WebSocket messages from Discord.
func (r *discordRPC) handleWebSocketMessage(connectionID, message string) error {
	if len(message) < 1024 {
		pdk.Log(pdk.LogTrace, fmt.Sprintf("Received WebSocket message for connection '%s': %s", connectionID, message))
	} else {
		pdk.Log(pdk.LogTrace, fmt.Sprintf("Received WebSocket message for connection '%s' (truncated): %s...", connectionID, message[:1021]))
	}

	// Parse the message
	var msg map[string]any
	if err := json.Unmarshal([]byte(message), &msg); err != nil {
		return fmt.Errorf("failed to parse WebSocket message: %w", err)
	}

	// Store sequence number if present
	if v := msg["s"]; v != nil {
		seq := int64(v.(float64))
		pdk.Log(pdk.LogTrace, fmt.Sprintf("Received sequence number for connection '%s': %d", connectionID, seq))
		if err := host.CacheSetInt(fmt.Sprintf("discord.seq.%s", connectionID), seq, int64(heartbeatInterval*2)); err != nil {
			return fmt.Errorf("failed to store sequence number for user %s: %w", connectionID, err)
		}
	}
	return nil
}

// handleHeartbeatCallback processes heartbeat scheduler callbacks.
func (r *discordRPC) handleHeartbeatCallback(username string) error {
	if err := r.sendHeartbeat(username); err != nil {
		// On first heartbeat failure, immediately clean up the connection
		pdk.Log(pdk.LogWarn, fmt.Sprintf("Heartbeat failed for user %s, cleaning up connection: %v", username, err))
		r.cleanupFailedConnection(username)
		return fmt.Errorf("heartbeat failed, connection cleaned up: %w", err)
	}
	return nil
}

// handleClearActivityCallback processes clear activity scheduler callbacks.
func (r *discordRPC) handleClearActivityCallback(username string) error {
	pdk.Log(pdk.LogInfo, fmt.Sprintf("Removing presence for user %s", username))
	if err := r.clearActivity(username); err != nil {
		return fmt.Errorf("failed to clear activity: %w", err)
	}

	pdk.Log(pdk.LogInfo, fmt.Sprintf("Disconnecting user %s", username))
	if err := r.disconnect(username); err != nil {
		return fmt.Errorf("failed to disconnect from Discord: %w", err)
	}
	return nil
}

File: plugins/examples/discord-rich-presence-rs/src/rpc.rs
//! Discord Rich Presence Plugin - RPC Communication
//!
//! This module handles all Discord gateway communication including WebSocket connections,
//! presence updates, and heartbeat management.

use extism_pdk::*;
use nd_pdk::host::{cache, scheduler, websocket};
use serde::{Deserialize, Serialize};

// ============================================================================
// Constants
// ============================================================================

const HEARTBEAT_OP_CODE: i32 = 1;
const GATE_OP_CODE: i32 = 2;
const PRESENCE_OP_CODE: i32 = 3;
const HEARTBEAT_INTERVAL: i32 = 41;
const DEFAULT_IMAGE: &str = "https://i.imgur.com/hb3XPzA.png";

const PAYLOAD_HEARTBEAT: &str = "heartbeat";

// ============================================================================
// Discord Types
// ============================================================================

#[derive(Serialize)]
pub struct Activity {
    pub name: String,
    #[serde(rename = "type")]
    pub activity_type: i32,
    pub details: String,
    pub state: String,
    #[serde(rename = "application_id")]
    pub application: String,
    pub timestamps: ActivityTimestamps,
    pub assets: ActivityAssets,
}

#[derive(Serialize)]
pub struct ActivityTimestamps {
    pub start: i64,
    pub end: i64,
}

#[derive(Serialize)]
pub struct ActivityAssets {
    pub large_image: String,
    pub large_text: String,
}

#[derive(Serialize)]
struct PresencePayload {
    activities: Vec<Activity>,
    since: i64,
    status: String,
    afk: bool,
}

#[derive(Serialize)]
struct IdentifyPayload {
    token: String,
    intents: i32,
    properties: IdentifyProperties,
}

#[derive(Serialize)]
struct IdentifyProperties {
    os: String,
    browser: String,
    device: String,
}

#[derive(Serialize)]
struct GatewayMessage<T> {
    op: i32,
    d: T,
}

#[derive(Deserialize)]
struct GatewayResponse {
    op: i32,
    #[serde(default)]
    #[allow(dead_code)]
    d: Option<serde_json::Value>,
    #[serde(default)]
    s: Option<i64>,
}

// ============================================================================
// Cache Keys
// ============================================================================

fn connection_key(username: &str) -> String {
    format!("discord.connection.{}", username)
}

fn token_key(username: &str) -> String {
    format!("discord.token.{}", username)
}

fn sequence_key(username: &str) -> String {
    format!("discord.sequence.{}", username)
}

// ============================================================================
// Connection Management
// ============================================================================

/// Tests if the connection is still valid by trying to send a heartbeat.
fn is_connected(username: &str) -> bool {
    match send_heartbeat(username) {
        Ok(_) => true,
        Err(e) => {
            trace!("Connection test failed for user {}: {:?}", username, e);
            false
        }
    }
}

/// Cleans up a connection for a user.
/// Called when heartbeat fails or connection is lost.
pub fn cleanup_connection(username: &str) {
    info!("Cleaning up failed connection for user {}", username);
    
    // Cancel the heartbeat schedule
    if let Err(e) = scheduler::cancel_schedule(username) {
        warn!("Failed to cancel heartbeat schedule for user {}: {:?}", username, e);
    }
    
    // Try to close the WebSocket connection
    let conn_key = connection_key(username);
    if let Ok(Some(conn_id)) = cache::get_string(&conn_key) {
        if !conn_id.is_empty() {
            if let Err(e) = websocket::close_connection(&conn_id, 1000, "Reconnecting") {
                trace!("Failed to close WebSocket for user {}: {:?}", username, e);
            }
            // Clean up reverse mapping
            let reverse_key = format!("discord.reverse.{}", conn_id);
            let _ = cache::remove(&reverse_key);
        }
    }
    
    // Clean up cache entries
    let _ = cache::remove(&conn_key);
    let _ = cache::remove(&sequence_key(username));
    
    info!("Cleaned up connection for user {}", username);
}

/// Handles connection close by connection ID (called from WebSocket close callback).
/// This cleans up all state associated with the connection.
pub fn handle_connection_close(connection_id: &str) {
    // Find the username for this connection using the reverse mapping
    if let Ok(Some(username)) = find_username_for_connection(connection_id) {
        info!("Connection closed for user {}, cleaning up", username);
        
        // Cancel the heartbeat schedule
        if let Err(e) = scheduler::cancel_schedule(&username) {
            // Not an error if schedule doesn't exist
            trace!("Failed to cancel heartbeat schedule for user {}: {:?}", username, e);
        }
        
        // Cancel any pending clear-activity schedule
        let _ = scheduler::cancel_schedule(&format!("{}-clear", username));
        
        // Clean up cache entries
        let conn_key = connection_key(&username);
        let _ = cache::remove(&conn_key);
        let _ = cache::remove(&sequence_key(&username));
        
        // Clean up reverse mapping
        let reverse_key = format!("discord.reverse.{}", connection_id);
        let _ = cache::remove(&reverse_key);
        
        info!("Cleaned up connection state for user {}", username);
    } else {
        // Just clean up the reverse mapping if we can't find the username
        let reverse_key = format!("discord.reverse.{}", connection_id);
        let _ = cache::remove(&reverse_key);
    }
}

/// Connects to the Discord gateway for a user.
pub fn connect(username: &str, token: &str) -> Result<(), Error> {
    // Check if already connected and connection is valid
    if is_connected(username) {
        info!("Reusing existing connection for user {}", username);
        return Ok(());
    }
    
    // Clean up any stale connection state
    cleanup_connection(username);

    info!("Connecting to Discord gateway for user {}", username);

    // Store token for later use
    cache::set_string(&token_key(username), token, 86400)?;

    // Get Discord Gateway URL
    let gateway = get_discord_gateway()?;
    info!("Using gateway: {}", gateway);

    // Connect to Discord gateway
    let headers = std::collections::HashMap::new();
    let conn_id = websocket::connect(
        &gateway,
        headers,
        username, // Use username as connection ID for easy lookup
    )?;
    info!("WebSocket connection established: {}", conn_id);

    // Store connection ID
    let conn_key = connection_key(username);
    cache::set_string(&conn_key, &conn_id, 86400)?;

    // Send identify immediately (don't wait for Hello)
    identify(username)?;

    info!("Successfully connected and identified user {}", username);
    Ok(())
}

/// Handles a WebSocket message from Discord.
pub fn handle_websocket_message(connection_id: &str, message: &str) -> Result<(), Error> {
    let response: GatewayResponse = serde_json::from_str(message)
        .map_err(|e| Error::msg(format!("Failed to parse gateway message: {}", e)))?;

    // Update sequence number if present
    if let Some(seq) = response.s {
        // Find username for this connection
        if let Some(username) = find_username_for_connection(connection_id)? {
            cache::set_string(&sequence_key(&username), &seq.to_string(), 86400)?;
        }
    }

    match response.op {
        10 => {
            // Hello - we already identified in connect(), nothing to do
        }
        11 => {
            // Heartbeat ACK - no action needed
        }
        1 => {
            // Heartbeat request - send heartbeat
            if let Some(username) = find_username_for_connection(connection_id)? {
                send_heartbeat(&username)?;
            }
        }
        _ => {
            trace!("Received Discord gateway op: {}", response.op);
        }
    }

    Ok(())
}

/// Handles heartbeat callback from scheduler.
pub fn handle_heartbeat_callback(username: &str) -> Result<(), Error> {
    send_heartbeat(username)
}

/// Handles clear activity callback from scheduler.
pub fn handle_clear_activity_callback(username: &str) -> Result<(), Error> {
    info!("Clearing activity for user {}", username);

    let conn_key = connection_key(username);
    if let Some(conn_id) = cache::get_string(&conn_key)?.filter(|s| !s.is_empty()) {
        // Send empty presence to clear activity
        let msg = GatewayMessage {
            op: PRESENCE_OP_CODE,
            d: PresencePayload {
                activities: vec![],
                since: 0,
                status: "dnd".to_string(),
                afk: false,
            },
        };

        let json = serde_json::to_string(&msg)
            .map_err(|e| Error::msg(format!("Failed to serialize message: {}", e)))?;

        websocket::send_text(&conn_id, &json)?;
    }

    Ok(())
}

/// Disconnects from Discord for a user.
pub fn disconnect(username: &str) -> Result<(), Error> {
    info!("Disconnecting from Discord for user {}", username);
    
    // Cancel the heartbeat schedule
    if let Err(e) = scheduler::cancel_schedule(username) {
        warn!("Failed to cancel heartbeat schedule: {:?}", e);
    }
    
    // Close the WebSocket connection
    let conn_key = connection_key(username);
    if let Some(conn_id) = cache::get_string(&conn_key)?.filter(|s| !s.is_empty()) {
        if let Err(e) = websocket::close_connection(&conn_id, 1000, "Navidrome disconnect") {
            warn!("Failed to close WebSocket connection: {:?}", e);
        }
        // Clean up reverse mapping
        let reverse_key = format!("discord.reverse.{}", conn_id);
        let _ = cache::remove(&reverse_key);
    }
    
    // Clean up cache entries
    let _ = cache::remove(&conn_key);
    let _ = cache::remove(&sequence_key(username));
    
    Ok(())
}

/// Sends an activity update to Discord.
pub fn send_activity(
    client_id: &str,
    username: &str,
    token: &str,
    mut activity: Activity,
) -> Result<(), Error> {
    let conn_key = connection_key(username);
    let conn_id = cache::get_string(&conn_key)?
        .filter(|s| !s.is_empty())
        .ok_or_else(|| Error::msg("Not connected to Discord"))?;

    // Process image URL
    activity.assets.large_image = process_image(&activity.assets.large_image, client_id, token)?;

    // Send presence update
    let msg = GatewayMessage {
        op: PRESENCE_OP_CODE,
        d: PresencePayload {
            activities: vec![activity],
            since: 0,
            status: "dnd".to_string(),
            afk: false,
        },
    };

    let json = serde_json::to_string(&msg)
        .map_err(|e| Error::msg(format!("Failed to serialize message: {}", e)))?;

    websocket::send_text(&conn_id, &json)?;

    Ok(())
}

// ============================================================================
// Internal Functions
// ============================================================================

fn find_username_for_connection(connection_id: &str) -> Result<Option<String>, Error> {
    // This is a simple approach - in production you might want to maintain a proper mapping
    // For now, we'll use a known pattern to find the username
    // The connection ID is stored as cache value, so we need to scan for it
    // Since we can't iterate cache, we'll use a workaround with a reverse mapping
    let reverse_key = format!("discord.reverse.{}", connection_id);
    Ok(cache::get_string(&reverse_key)?.filter(|s| !s.is_empty()))
}

fn get_discord_gateway() -> Result<String, Error> {
    let req = HttpRequest::new("https://discord.com/api/gateway")
        .with_method("GET");

    let resp = http::request::<String>(&req, None::<String>)?;
    if resp.status_code() >= 400 {
        return Err(Error::msg(format!(
            "Failed to get Discord gateway: HTTP {}",
            resp.status_code()
        )));
    }

    let body = resp.body();
    let data: std::collections::HashMap<String, String> = serde_json::from_slice(&body)
        .map_err(|e| Error::msg(format!("Failed to parse gateway response: {}", e)))?;

    data.get("url")
        .map(|url| url.to_string())
        .ok_or_else(|| Error::msg("No URL in gateway response"))
}

fn identify(username: &str) -> Result<(), Error> {
    info!("Identifying with Discord for user {}", username);

    let conn_key = connection_key(username);
    let conn_id = cache::get_string(&conn_key)?
        .filter(|s| !s.is_empty())
        .ok_or_else(|| Error::msg("No connection found"))?;

    let token_k = token_key(username);
    let token = cache::get_string(&token_k)?
        .filter(|s| !s.is_empty())
        .ok_or_else(|| Error::msg("No token found"))?;

    // Store reverse mapping for connection -> username
    let reverse_key = format!("discord.reverse.{}", conn_id);
    cache::set_string(&reverse_key, username, 86400)?;

    // Send identify
    let msg = GatewayMessage {
        op: GATE_OP_CODE,
        d: IdentifyPayload {
            token,
            intents: 0,
            properties: IdentifyProperties {
                os: "Windows 10".to_string(),
                browser: "Discord Client".to_string(),
                device: "Discord Client".to_string(),
            },
        },
    };

    let json = serde_json::to_string(&msg)
        .map_err(|e| Error::msg(format!("Failed to serialize message: {}", e)))?;

    websocket::send_text(&conn_id, &json)?;

    // Schedule heartbeat
    scheduler::schedule_recurring(
        &format!("@every {}s", HEARTBEAT_INTERVAL),
        PAYLOAD_HEARTBEAT,
        username,
    )?;

    Ok(())
}

fn send_heartbeat(username: &str) -> Result<(), Error> {
    let conn_key = connection_key(username);
    let conn_id = cache::get_string(&conn_key)?
        .filter(|s| !s.is_empty())
        .ok_or_else(|| Error::msg("No connection found"))?;

    // Get sequence number
    let seq_key = sequence_key(username);
    let seq: Option<i64> = cache::get_string(&seq_key)?
        .and_then(|s| s.parse().ok());

    // Send heartbeat
    let msg = GatewayMessage {
        op: HEARTBEAT_OP_CODE,
        d: seq,
    };

    let json = serde_json::to_string(&msg)
        .map_err(|e| Error::msg(format!("Failed to serialize message: {}", e)))?;

    websocket::send_text(&conn_id, &json)?;
    Ok(())
}

fn process_image(image_url: &str, client_id: &str, token: &str) -> Result<String, Error> {
    process_image_inner(image_url, client_id, token, false)
}

fn process_image_inner(
    image_url: &str,
    client_id: &str,
    token: &str,
    is_default: bool,
) -> Result<String, Error> {
    let url = if image_url.is_empty() {
        if is_default {
            return Err(Error::msg("default image URL is empty"));
        }
        return process_image_inner(DEFAULT_IMAGE, client_id, token, true);
    } else {
        image_url
    };

    // Already processed
    if url.starts_with("mp:") {
        return Ok(url.to_string());
    }

    // Check cache
    let cache_key = format!("discord.image.{:x}", md5_hash(url));
    if let Some(cached) = cache::get_string(&cache_key)?.filter(|s| !s.is_empty()) {
        return Ok(cached);
    }

    // Process via Discord API
    let body = format!(r#"{{"urls":["{}"]}}"#, url);
    let api_url = format!(
        "https://discord.com/api/v9/applications/{}/external-assets",
        client_id
    );

    let req = HttpRequest::new(&api_url)
        .with_method("POST")
        .with_header("Authorization", token)
        .with_header("Content-Type", "application/json");

    let resp = http::request::<String>(&req, Some(body))?;
    if resp.status_code() >= 400 {
        if is_default {
            return Err(Error::msg(format!(
                "failed to process default image: HTTP {}",
                resp.status_code()
            )));
        }
        return process_image_inner(DEFAULT_IMAGE, client_id, token, true);
    }

    let body = resp.body();
    let data: Vec<std::collections::HashMap<String, String>> = serde_json::from_slice(&body)
        .map_err(|e| Error::msg(format!("Failed to parse image response: {}", e)))?;

    if data.is_empty() {
        if is_default {
            return Err(Error::msg("no data returned for default image"));
        }
        return process_image_inner(DEFAULT_IMAGE, client_id, token, true);
    }

    let asset_path = data[0]
        .get("external_asset_path")
        .map(|s| s.as_str())
        .unwrap_or("");

    if asset_path.is_empty() {
        if is_default {
            return Err(Error::msg("empty external_asset_path for default image"));
        }
        return process_image_inner(DEFAULT_IMAGE, client_id, token, true);
    }

    let processed = format!("mp:{}", asset_path);

    // Cache the result
    let ttl = if is_default { 48 * 60 * 60 } else { 4 * 60 * 60 };
    let _ = cache::set_string(&cache_key, &processed, ttl);

    Ok(processed)
}

/// Simple hash function for cache keys.
fn md5_hash(input: &str) -> u64 {
    // A simple hash - not actual MD5, but sufficient for cache keys
    let mut hash: u64 = 0;
    for (i, byte) in input.bytes().enumerate() {
        hash = hash.wrapping_add((byte as u64).wrapping_mul((i as u64).wrapping_add(1)));
        hash = hash.wrapping_mul(31);
    }
    hash
}
